### PR TITLE
fix(search_family): Process wrong field types in indexes for the FT.SEARCH and FT.AGGREGATE commands

### PR DESCRIPTION
**Pull request by BagritsevichStepan**
fixes dragonflydb#3986

### CODE REVIEW

File: src/core/search/base.cc
Comment (romange) [inline src/core/search/base.cc:44]: ` || value.empty()` looks like a bug

File: src/core/search/indices.cc
Comment (romange) [inline src/core/search/indices.cc:82]: the assumption here is that all values are convertible to double? why is that?

File: src/core/search/indices.cc
Comment (romange) [inline src/core/search/indices.cc:76]: so here you test that all field values are indexable? why all of them?

File: src/core/search/search.h
Comment (romange) [inline src/core/search/search.h:80]: please add a comment what it returns

File: src/core/search/sort_indices.cc
Comment (romange) [inline src/core/search/sort_indices.cc:76]: `str` is ambigous clear. it is a vector? either name it `str_vec` or provide an explicit type

File: src/core/search/sort_indices.cc
Comment (romange) [inline src/core/search/sort_indices.cc:82]: why it's a warning? can it happen as a result of user input? is it a bug or a valid scenario?

File: src/core/search/sort_indices.cc
Comment (romange) [inline src/core/search/sort_indices.cc:91]: same, lets call it `str_vec` or `str_arr`

File: src/server/search/doc_accessors.cc
Comment (romange) [inline src/server/search/doc_accessors.cc:47]: this will throw imho if `value_as_double` is nullopt, so it's better use CHECK fail  above.

**Review by romange — COMMENTED** [review]
I will wait for @dranikpg  to review , meanwhile some minor comments so I could understand the PR better

File: src/core/search/sort_indices.cc
Comment (BagritsevichStepan) [reply src/core/search/sort_indices.cc:82]: Yes, it can happen as a result of user input

File: src/core/search/indices.cc
Comment (BagritsevichStepan) [reply src/core/search/indices.cc:76]: Actually the result of GetStrings should be only one value

File: src/core/search/sort_indices.cc
Comment (BagritsevichStepan) [reply src/core/search/sort_indices.cc:82]: I think it would be better to remove this `LOG`

File: src/core/search/indices.cc
Comment (BagritsevichStepan) [reply src/core/search/indices.cc:76]: From Redis:
```
127.0.0.1:6379> json.set j1 . '{"arr":[{"value":1},{"value":"one"}]}'
OK
127.0.0.1:6379> json.set j2 . '{"arr":[{"value":1},{"value":2}]}'
OK
127.0.0.1:6379> ft.create index on json schema $.arr[*].value AS value NUMERIC
OK
127.0.0.1:6379> ft.search index "*"
1) (integer) 1
2) "j2"
3) 1) "$"
   2) "{\"arr\":[{\"value\":1},{\"value\":2}]}"
```

Comment (BagritsevichStepan) [reply src/core/search/indices.cc:76]: So all fields should have the same field type as index. Only then this document can be added

File: src/core/search/indices.cc
Comment (BagritsevichStepan) [reply src/core/search/indices.cc:82]: Before it was matched => proved that all fields are doubles

File: src/core/search/indices.cc
Comment (BagritsevichStepan) [reply src/core/search/indices.cc:76]: I also see that we had another logic for the sort indices. We just parsed only the first value. It is a bug. But there is another problem that we are saving the sortable values

File: src/core/search/base.cc
Comment (BagritsevichStepan) [reply src/core/search/base.cc:44]: Why?

From Redis:
```
127.0.0.1:6379> json.set j1 . '{"value":1}'
OK
127.0.0.1:6379> json.set j2 . '{"another_value":"not number"}'
OK
127.0.0.1:6379> ft.create index on json schema $.value as value numeric
OK
127.0.0.1:6379> ft.search index "*"
1) (integer) 2
2) "j2"
3) 1) "$"
   2) "{\"another_value\":\"not number\"}"
4) "j1"
5) 1) "$"
   2) "{\"value\":1}"
```

So here Redis indexes `j2`, even if it doesn't have the required field (i.e., the field value is an empty string).

Comment (BagritsevichStepan) [reply src/core/search/base.cc:44]: Yes, I found the edge case: `json.set j1 . '{"value":""}'`

**Comment by BagritsevichStepan** [comment]
@romange I've added more tests. After adding support for nullable values, a more tests will be added. Please take a look at the PR.

**Review by dranikpg — COMMENTED** [review]
My two cents:
1. Please write next time what you changed in the PR description, so I can dive into the changes quicker
2. Maybe look into unifying Matches() and Add(), this would simplify code
3. Please check vectors or add a TODO

If we hurry for a release I can approve 🙂

File: src/core/search/indices.cc
Comment (BagritsevichStepan) [reply src/core/search/indices.cc]: Yes, I added checking and tests

File: src/core/search/search.cc
Comment (BagritsevichStepan) [reply src/core/search/search.cc]: Yes, you are right

File: src/core/search/base.h
Comment (BagritsevichStepan) [reply src/core/search/base.h]: I understand that we could implement it as `bool Add()`. I did it this way because, by the next release, we might be able to implement `std::optional<Data> Matches()`, with `Add` taking this data as an argument. This way, we won’t need to do a rollback, and we’ll fetch the data only once. Moreover, we already have a cache in `JsonAccessor`.

**Comment by BagritsevichStepan** [comment]
I found several new bugs and fixed them in the latest commit:
1. Bug:
```
JSON.SET j1 . '{"data":1}'
JSON.SET j2 . '{"data":"1"}'
FT.CREATE i1 ON JSON SCHEMA $.data AS data NUMERIC
FT.CREATE i2 ON JSON SCHEMA $.data AS data TEXT
```
`FT.SEARCH i1 "*"` and `FT.SEARCH i2 "*"` both return j1 and j2. However, it should return j1 for i1 and j2 for i2. This happens because we are parsing string and numeric fields in the same way in the `DocumentAccessor`.

2. Bug related to how we parse `FtVector` in the `GetVector` function in `DocumentAccessor`.

I’ve also added more tests

**Comment by BagritsevichStepan** [comment]
Found and fixed another bug:

We should support queries like this:
```
JSON.SET j1 . '{"data":["first value", "second_value"]}'
FT.CREATE index ON JSON SCHEMA $.data AS data TEXT
```
`FT.SEARCH index "*"` should return `j1`

File: src/core/search/base.h
Comment (romange) [inline src/core/search/base.h:93]: nit: please use `const Type&` wherever possible.

File: src/core/search/base.h
Comment (romange) [inline src/core/search/base.h:106]:
```suggestion
/* Used for converting field values to double. Returns nullopt if the conversion fails */
```

File: src/server/search/doc_accessors.h
Comment (romange) [inline src/server/search/doc_accessors.h:42]: please move the implementation to cc file

File: src/core/search/base.h
Comment (dranikpg) [inline src/core/search/base.h:71]: not shorter than std::optional and confuses the reader with a new type 😅

**Review by dranikpg — DISMISSED** [review]
Good job fixing bugs around vector parsing, it wasn't safe before. 

Still, once we merge this for 1.25, I'd suggest to unite IsValidFieldType and Add, because we have safe/unsafe parsing functions, optionals that are dereferenced "by implicit contract" and double effort indexing documents (we parse it two times)

File: src/server/search/doc_accessors.h
Comment (dranikpg) [reply src/server/search/doc_accessors.h:42]: Oh, yes, and this

**Review by romange — APPROVED** [review]

**Pull request merged**
